test/watch.js

```diff
@@ -2,43 +2,144 @@ const test = require('brittle')
 const { create, createRange, createStored, eventFlush } = require('./helpers')
 
 test('basic watch', async function (t) {
+  t.plan(3)
+
+  const db = create()
+  const watcher = db.watch()
+  t.teardown(() => watcher.destroy())
+
+  db.put('/a') // Run on background
+
+  const { done, value: { current, previous } } = await watcher.next()
+
+  t.is(done, false)
+  t.is(current.version, 2)
+  t.is(previous.version, 1)
+})
+
+test('watch multiple next() on parallel - value', async function (t) {
+  t.plan(9)
+
+  const db = create()
+  const watcher = db.watch()
+  t.teardown(() => watcher.destroy())
+
+  const a = watcher.next()
+  const b = watcher.next()
+  const c = watcher.next()
+
+  db.put('/a') // Run on background
+
+  {
+    const { done, value: { current, previous } } = await a
+
+    t.is(done, false)
+    t.is(current.version, 2)
+    t.is(previous.version, 1)
+  }
+
+  db.put('/b') // Run on background
+
+  {
+    const { done, value: { current, previous } } = await b
+
+    t.is(done, false)
+    t.is(current.version, 3)
+    t.is(previous.version, 2)
+  }
+
+  db.put('/c') // Run on background
+
+  {
+    const { done, value: { current, previous } } = await c
+
+    t.is(done, false)
+    t.is(current.version, 4)
+    t.is(previous.version, 3)
+  }
+})
+
+test('watch multiple next() on parallel - done', async function (t) {
   t.plan(2)
 
   const db = create()
+  const watcher = db.watch()
+
+  const a = watcher.next()
+  const b = watcher.next()
+
+  await watcher.destroy()
+
+  t.alike(await a, { done: true, value: undefined })
+  t.alike(await b, { done: true, value: undefined })
+})
+
+test('watch next() after is destroyed', async function (t) {
+  t.plan(1)
+
+  const db = create()
+  const watcher = db.watch()
+
+  await watcher.destroy()
+
+  t.alike(await watcher.next(), { done: true, value: undefined })
+})
+
+test('watch waits for new change', async function (t) {
+  t.plan(3)
+
+  const db = create()
+  await db.put('/a') // Ignore first append (header)
 
   const watcher = db.watch()
   t.teardown(() => watcher.destroy())
 
-  watcher.on('change', function (newVersion, oldVersion) {
-    t.is(newVersion, 2)
-    t.is(oldVersion, 1)
+  setImmediate(async () => {
+    await eventFlush()
+    db.put('/b') // Run on background
   })
```

Comment:
why the setImmediate? vs eventFlush().then(() => db.put(...))

Reply:
100% right

We need a `require('mafintosh')` that removes `setImmediate` globally so I don't keep doing the same mistake (btw NPM name available if you want to take it xD)